### ISSUE TITLE
s25client executable cannot be found

### ISSUE THREAD

**Issue by stefson** [Return-To-The-Roots/s25client#616]
I just ran into the problem that I cannot start the client anymore by typing s25client into a terminal, all I get is this error 

> Executable not at '"/home/user/s25client"'
> 
> Press ENTER to close this window . . .

the s25client binary is of course not in the home folder but to be found under /usr/bin/s25client, so how can this happen? These are self build binaries, and all which I know is that you have recently changed some internal stuff in the build system to bind to certain paths, is that correct? 

My system is a gentoo amd64, cmake 3.6, gcc 4.9.4 and glibc 2.22 if that matters. I am happy to test any kind of patches but have no idea how to solve this on my own and provide a patch for you, sry about that. However, the client starts as usual when typing /usr/bin/s25client instead of just s25client into the terminal.

**Comment by stefson** [comment]
Can someone please explain me where the value for the install prefix/prefixed binary is stored? I don't mind to adjust it locally for my needs, but can't get this sorted out on my own.

**Comment by stefson** [comment]
The binary s25client assumes it is put into the home folder, meaning the value for Configured binary dir: "bin" should be changed to /usr/bin/ to work properly. Right now it is guessing for the prefixed binary dir in: "/home/bin" 

full error, if I symlink the binary from /usr/bin/s25client to ~/s25client and try to launch it from the commandline is 

Could not find install prefix.
Current binary dir: "/home/tim"
Best guess for prefixed binary dir: "/home/bin"
Configured binary dir: "bin"


Press ENTER to close this window . . .

**Comment by Flamefire** [comment]
Problem is here: https://github.com/Return-To-The-Roots/s25client/blob/master/src/ProgramInitHelpers.cpp#L67
I have no idea ATM how to find out the actual path to the current exe file if the exe is in the PATH and not started with a full path. All that shows up in the program is then "s25client" and the best guess is that it is in the current directory which isn't the case if the s25client is in a path that is in PATH.

**Comment by Flamefire** [comment]
Changing something in #648 which should also fix this. In the meantime set the environment variable `RTTR_PREFIX_PATH` to `/usr` and it should work.
Please test after #648 is merged without the env var.

**Comment by stefson** [comment]
I will try to find out how to set this path on Gentoo and keep you posted on this. 

However, there is one thing I don't quiet understand: I get a cmake warning 

CMake Warning:
  Manually-specified variables were not used by the project:
    BUILD_GLFW
    COMPILEARCH
    COMPILEFOR
    RTTR_INSTALL_PREFIX 

Does this mean the code which I check out from here does not set those variables in the cmake skripts, or does it mean the Gentoo way to build cmake stuff does not set or honor and overwrite these variables?

**Comment by Flamefire** [comment]
It means that they are not USED. Got to check, why or how to suppress the warning. Some of them are used later or where used, maybe not anymore

**Comment by stefson** [comment]
From my understanding all RTTR_*DIR variables must be relative paths, and later on (at runtime?) they get resolved relative to  RTTR_BINDIR, is that correct? I mean this issue here can also be put on hold until you resolve #647 , it is not really urgent in my opinion.

**Comment by Flamefire** [comment]
No, they are relative to the prefix as even bindir is relative to that. Besides that you are right

**Comment by stefson** [comment]
Well, the gentoo build script uses local cmake arguments, one of them is -DRTTR_INSTALL_PREFIX=/usr 
These cmake arguments are applied during the configure phase, but if this has finished I get a warning about missing RTTR_INSTALL_PREFIX - nonetheless it compiles fine, just a bit broken :)

**Comment by Flamefire** [comment]
Then the script has to be updated. Those variables (all 4 of them) are not used anymore. Please note that the warning is NOT about a missing variable but that those are UNUSED which is correct. They were used before the rework of the install system.

**Comment by stefson** [comment]
#648 caused another issue und did not really fix the initial problem, at least on linux. 

The binary provided by you throws out the following error

Update finished!
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_S_construct null not valid
/home/tim/s25rttr_20150922/bin/../bin/rttr.sh: Zeile 60: 12264 Abgebrochen             $cmd $DIR/../bin/s25client $*
An error occured: press enter to continue

**Comment by stefson** [comment]
The binary I build on my own from the git sources (if I use /usr/bin/s25client to start, s25client does not work) 

terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid

backtrace from gdb 

Program received signal SIGABRT, Aborted.
0x00007ffff5fb8148 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54	../sysdeps/unix/sysv/linux/raise.c: Datei oder Verzeichnis nicht gefunden.
(gdb) bt
#0  0x00007ffff5fb8148 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff5fb959a in __GI_abort () at abort.c:89
#2  0x00007ffff66b079d in __gnu_cxx::__verbose_terminate_handler() () from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/libstdc++.so.6
#3  0x00007ffff66ae606 in ?? () from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/libstdc++.so.6
#4  0x00007ffff66ae651 in std::terminate() () from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/libstdc++.so.6
#5  0x00007ffff66ae868 in __cxa_throw () from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/libstdc++.so.6
#6  0x00007ffff66d723f in std::__throw_logic_error(char const*) () from /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/libstdc++.so.6
#7  0x0000000000807a34 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*> (this=this@entry=0x7fffffffd5c0, 
    __beg=__beg@entry=0x0, __end=<optimized out>) at /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/basic_string.tcc:216
#8  0x0000000000807b0f in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct_aux<char const*> (__end=<optimized out>, __beg=0x0, 
    this=0x7fffffffd5c0) at /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/basic_string.h:195
#9  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*> (__end=<optimized out>, __beg=0x0, this=0x7fffffffd5c0)
    at /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/basic_string.h:214
#10 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string (__a=..., __s=0x0, this=0x7fffffffd5c0)
    at /usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/bits/basic_string.h:456
#11 System::getEnvVar (name="RTTR_PREFIX_DIR") at /var/tmp/portage/games-strategy/s25rttr-9999-r2/work/s25rttr-9999/libutil/src/System.cpp:99
#12 0x0000000000807c04 in System::getPathFromEnvVar (name="RTTR_PREFIX_DIR") at /var/tmp/portage/games-strategy/s25rttr-9999-r2/work/s25rttr-9999/libutil/src/System.cpp:122
#13 0x00000000005acd92 in InitWorkingDirectory (exeFilepath=...) at /var/tmp/portage/games-strategy/s25rttr-9999-r2/work/s25rttr-9999/src/ProgramInitHelpers.cpp:84
#14 0x0000000000551160 in RunProgram (exeFilepath="/usr/bin/s25client", options=...) at /var/tmp/portage/games-strategy/s25rttr-9999-r2/work/s25rttr-9999/src/main.cpp:348
#15 0x0000000000504c06 in main (argc=1, argv=0x7fffffffdd88) at /var/tmp/portage/games-strategy/s25rttr-9999-r2/work/s25rttr-9999/src/main.cpp:429

**Comment by stefson** [comment]
Could it be that you forgot to lift up the version of libutil? I get an error of a missing System.h if I am compiling the source in the way you are suggesting with ./cmake.sh --prefix=. && make from the build directory. 

[buildlog.zip](https://github.com/Return-To-The-Roots/s25client/files/1116883/buildlog.zip)

**Comment by Flamefire** [comment]
Nope did not forgot it. However I made a mistake there where I don't check for a NULL pointer. Writing a testcase now.
About the missing include: try `git submodule update` after a `git pull`

**Comment by stefson** [comment]
Indeed, one has to update the submodules in the way of git submodule update rather than to do a git pull from within the libutil folder. The crash is fixed, but the initial problem is not, I still can't go via s25client but have to go via /usr/bin/s25client.

**Comment by Flamefire** [comment]
Interesting. Can you print out the values before and after the system_complete call?

**Comment by stefson** [comment]
I certainly could if I knew what system_complete call is and how I got these values?

**Comment by Flamefire** [comment]
See https://github.com/Return-To-The-Roots/s25client/blob/master/src/ProgramInitHelpers.cpp#L75
Just do something like `std::cout << exeFilepath << std::endl << bfs::system_complete(exeFilepath) << std::endl` before this line (untested) compile and report output please.

**Comment by stefson** [comment]
if I save this expression in test.cpp (it is c++, isn't it?) and do a 

gcc test.cpp -c on it, it fails with 

test.cpp:1:6: Fehler: »cout« in Namensbereich »std« bezeichnet keinen Typ 
std::cout << exeFilepath << std::endl << bfs::system_complete(exeFilepath) << 

can you work with that? I use dynamical boost-1.62 system libs and gcc 5.4.0

**Comment by Flamefire** [comment]
You have to put in in rttr of course. I linked the file and the line above then compile and run as usual

**Comment by stefson** [comment]
Ah, okay, I thought it was a test case to check for a potential compiler bug! So I made up a diff, is this what you ententioned me to apply? 
```

diff --git a/src/ProgramInitHelpers.cpp b/src/ProgramInitHelpers.cpp
index 59e0cb9..7945bb5 100644
--- a/src/ProgramInitHelpers.cpp
+++ b/src/ProgramInitHelpers.cpp
@@ -72,7 +72,7 @@ bool InitWorkingDirectory(const bfs::path& exeFilepath)
     // Complete the path as it would be done by the system
     // This avoids problems if the program was not started from the working directory
     // e.g. by putting its path in PATH
-    bfs::path fullExeFilepath = bfs::absolute(bfs::system_complete(exeFilepath));
+    std::cout << exeFilepath << std::endl << bfs::system_complete(exeFilepath) << std::endl
     if(!bfs::exists(fullExeFilepath))
     {
         std::cerr << "Executable not at '" << fullExeFilepath << "'" << std::endl;

```
If yes, than I can tell you that it breaks the compile: 

```
 /var/tmp/portage/games-strategy/s25rttr-9999-r2/work/s25rttr-9999/src/ProgramInitHelpers.cpp: In function ‘bool InitWorkingDirectory(const boost::filesystem::path&)’:
/var/tmp/portage/games-strategy/s25rttr-9999-r2/work/s25rttr-9999/src/ProgramInitHelpers.cpp:76:5: error: expected ‘;’ before ‘if’
     if(!bfs::exists(fullExeFilepath))
     ^
/var/tmp/portage/games-strategy/s25rttr-9999-r2/work/s25rttr-9999/src/ProgramInitHelpers.cpp:90:37: error: ‘fullExeFilepath’ was not declared in this scope
         const bfs::path curBinDir = fullExeFilepath.parent_path();

```

**Comment by Flow86** [comment]
you should not remove the line

````
bfs::path fullExeFilepath = bfs::absolute(bfs::system_complete(exeFilepath));
````

**Comment by stefson** [comment]
please send me a diff, I can't read code like you guys

**Comment by Flow86** [comment]
```

diff --git a/src/ProgramInitHelpers.cpp b/src/ProgramInitHelpers.cpp
index 59e0cb9..7945bb5 100644
--- a/src/ProgramInitHelpers.cpp
+++ b/src/ProgramInitHelpers.cpp
@@ -72,7 +72,8 @@ bool InitWorkingDirectory(const bfs::path& exeFilepath)
     // Complete the path as it would be done by the system
     // This avoids problems if the program was not started from the working directory
     // e.g. by putting its path in PATH
     bfs::path fullExeFilepath = bfs::absolute(bfs::system_complete(exeFilepath));
+    std::cout << exeFilepath << std::endl << bfs::system_complete(exeFilepath) << std::endl
     if(!bfs::exists(fullExeFilepath))
     {
         std::cerr << "Executable not at '" << fullExeFilepath << "'" << std::endl;

```

**Comment by stefson** [comment]
thanks for the diff! 
but the compilation is broken: 

```/var/tmp/portage/games-strategy/s25rttr-9999-r2/work/s25rttr-9999/src/ProgramInitHelpers.cpp: In function ‘bool InitWorkingDirectory(const boost::filesystem::path&)’:
/var/tmp/portage/games-strategy/s25rttr-9999-r2/work/s25rttr-9999/src/ProgramInitHelpers.cpp:77:5: error: expected ‘;’ before ‘if’
     if(!bfs::exists(fullExeFilepath))
     ^
make[2]: *** [src/CMakeFiles/s25Main.dir/build.make:1048: src/CMakeFiles/s25Main.dir/ProgramInitHelpers.cpp.o] Error 1
```

**Comment by Flamefire** [comment]
Don't worry for now. The current solution still won't work. Will implement something else.
Dev note: See https://stackoverflow.com/questions/1528298/get-path-of-executable and https://stackoverflow.com/questions/1023306/finding-current-executables-path-without-proc-self-exe

**Comment by Flow86** [comment]
oh there's a ";" missing at the end

**Comment by stefson** [comment]
With the ";" it compiles, but there is no effect. Game can be started via /usr/bin/s25client and runs as expected, but not via s25client.

**Comment by stefson** [comment]
this solved the problem, thank you! :-)
